Sources/Logging/Logging.swift

```diff
@@ -144,20 +179,15 @@ public final class StdoutLogger: LogHandler {
         }
     }
 
-    private var prettyMetadata: String?
-    private var _metadata: LoggingMetadata? {
-        didSet {
-            self.prettyMetadata = !(self._metadata?.isEmpty ?? true) ? self._metadata!.map { "\($0)=\($1)" }.joined(separator: " ") : nil
-        }
-    }
-
+    private var _metadata: LoggingMetadata
     public func log(level: LogLevel, message: String, file _: String, function _: String, line _: UInt) {
         if level >= self.logLevel {
-            print("\(Date()) \(level)\(self.prettyMetadata.map { " \($0)" } ?? "") \(message)")
+            let prettyMetadata = !(self._metadata.isEmpty) ? self._metadata.map { "\($0)=\($1)" }.joined(separator: " ") : ""
```

Comment:
Right, I agree that "rendering many times" is wasteful, however so is "rendering when we don't ever print" the pretty metadata, yet the field is changed a lot.

So the downsides of the 2 styles are:

- `prettyMetadata` String property, re-rendered at `metadata didSet` – we render all the values even if they end up not used. This is specifically hurting in the case of: used log level determines we would not log things, we set metadata per each execution, it is different because depends on the data (e.g. a trace's parent id), we render the prettyMetadata even though we don't know if we'll print it. 
  - a) we don't know because maybe there is no `log` call at all 
  - b) we don't know* because based on loglevel just now, maybe it should be printed... but then loglevel is changed, and maybe we now would print or now maybe we would not. So we'd have to inspect log level here, but also potentially trigger rendering when loglevel is changed. but even then we may render needlessly, as loglevel was changed, but nothing was logged – so we did a wasteful render of the `prettyMetadata` again.
- the "in this PR" proposal, renders only when we log – however it renders many many times, for each log call. This is also wasteful, if the logging never changed.

Note on hot path: I may be biased by my execution styles I work with, but the setting of the metadata is immediately right before the code executing user code, so it does not change the entire "time to process". It would change the "time spent in user handler" though, if we render before yes... I'm not convienced this matters in the end, as users look at "what's my throughput", and both executions happen in the same hot loop in either case.

I think we can achieve best of both worlds, perhaps. How about:

- keep the `prettyMetadata: String?` property
- in `metadata didSet { prettyMetadata = nil }`, so we invalidated our cached `prettyMetadata` since the values changed.
- user code runs:
  - case a) no logging happens, we never render this specific set of metadata (good)
  - case b) logging happens, on log level that does not trigger actual printing though – no rendering happens (good)
  - case c) logging happens, on log level that causes printing; if the `prettyMetadata` is `nil` we render once and initialize it; subsequent logging reuses the once rendered metadata – we avoided double rendering, and avoided rendering too early (good, in my book at least)
- some cases to consider when loglevel is changed in user code:
  - case a) log event happens, though it is `trace` and we only log from `info`, meaning we did not render the metadata, user code sets the log level to `trace` so we should now render it; the next log on `trace` causes the render to happen and caches it (good)
  - case b) user logs at various log levels, if at least one of them has to be printed we render; we don't render twice though (good)

So we get the cached `prettyMetadata` and also avoid rendering it in case no log event would make use of it.

WDYT?

Reply:
(Note to self: so technically this is an implementation thing, however at the same time it relates to if/how we restrict the type of values to be Strings in the Metadata or not; if not, then these optimizations come into play; if they are forced to be strings already it matters less, as we were forced to make the rendering before already...)

// I'm looking at context APIs now and if/how they could allow avoiding these issues, though don't have an answer yet.